### ISSUE TITLE
Improve threadsafety of concurrent calls and (implicit or explicit) configuration changes to Fakes 

### ISSUE THREAD

**Issue by RodneyRichardson** [FakeItEasy/FakeItEasy#1569]
I have seen the following exception in my multi-threaded tests. 

```
System.InvalidOperationException: Collection was modified after the enumerator was instantiated.
   at System.Collections.Generic.LinkedList`1.Enumerator.MoveNext()
   at FakeItEasy.Core.FakeManager.ApplyBestRule(IInterceptedFakeObjectCall fakeObjectCall) in C:\projects\fakeiteasy\src\FakeItEasy\Core\FakeManager.cs:line 210
   at FakeItEasy.Core.FakeManager.FakeItEasy.Core.IFakeCallProcessor.Process(InterceptedFakeObjectCall fakeObjectCall) in C:\projects\fakeiteasy\src\FakeItEasy\Core\FakeManager.cs:line 136
   at Castle.DynamicProxy.AbstractInvocation.Proceed() in C:\projects\fakeiteasy\src\FakeItEasy\Expressions\ExpressionArgumentConstraintFactory.cs:line 0
   at Castle.Proxies.ObjectProxy_3.get_MyProperty()
   ... {my code removed - starts at the handler for MyEvent}
```

I have the following scenario:
Code in my test:
```c#
A.CallTo(() => m_MyObject.MyProperty).Returns(SomeValue);
m_MyObject.MyEvent += Raise.With(new MyEventArgs());
A.CallTo(() => m_MyObject.MyProperty).Returns(SomeOtherValue);
```

The event is handled by my code, marshaled to another thread, which reads the value of MyProperty. I think there's a thread issue with setting the return value of MyProperty at the same time as it's being read on another thread.

**Comment by blairconrad** [comment]
@RodneyRichardson, I'm looking at the test code, and I have concerns about it. Regardless of whether we choose to make a change to `ApplyBestRule`, I think it's unsafe.

If your eventhandler is reading the `MyProperty` value on another thread, and there's no synchronization between the threads, you're not guaranteed of the value that will be returned. `A.CallTo(() => m_MyObject.MyProperty).Returns(SomeOtherValue)` might be called _before_ your eventhandler reads `MyProperty`.

**Comment by blairconrad** [comment]
I think we have a bug. On the one hand, as I noted earlier, I don't think the presented test is in its best form, and I would generally encourage testwriters to establish Fakes' configuration _before_ exercising the fakes, but I think there are still some opportunities for a conflict to happen even when this rule is heeded.
For example, `PropertySetterRule` and `AutoFakePropertyRule` may both mutate the `AllUserRules` list when a Fake is _used_, not configured, and could do so while other threads are iterating over the list.

I think it's worthwhile plugging those holes, even if we choose not to support the kind of explicit Fake-configuring presented in this issue's description.

As an aside, I think there's another opportunity for multithreaded Fake calls to conflict: `ReturnsNextFromSequence` is implemented by constructing a `Queue` and adding a rule that dequeues values from it. If two threads trigger the action concurrently, they may behave badly.

**Comment by thomaslevesque** [comment]
Yes, we should protect against concurrent access from different threads. A straightforward approach would be to put a lock around any access to `AllUserRules` (but release the lock before applying the rule). It might negatively affect performance, though...

**Comment by blairconrad** [comment]
Yeah, I had the same thoughts. (And a `ConcurrentQueue` for `ReturnsNextFromSequence`? Or we could just add many many individual rules. :wink:

**Comment by blairconrad** [comment]
I'm happy to work on this issue.

**Comment by thomaslevesque** [comment]
> I'm happy to work on this issue.

👍

**Comment by RodneyRichardson** [comment]
> @RodneyRichardson, I'm looking at the test code, and I have concerns about it. Regardless of whether we choose to make a change to `ApplyBestRule`, I think it's unsafe.
> 
> If your eventhandler is reading the `MyProperty` value on another thread, and there's no synchronization between the threads, you're not guaranteed of the value that will be returned. `A.CallTo(() => m_MyObject.MyProperty).Returns(SomeOtherValue)` might be called _before_ your eventhandler reads `MyProperty`.

Yes, I agree. In my real code, the real object which implements MyProperty and MyEvent is threadsafe.

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 5.1.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/5.1.0).


Thanks, @RodneyRichardson. Look for your name in the release notes. :trophy:
